src/Spanner/Session/CacheSessionPool.php

```diff
@@ -0,0 +1,548 @@
+ * use Symfony\Component\Cache\Adapter\FilesystemAdapter;
+ *
+ * $spanner = new SpannerClient();
+ * $cache = new FilesystemAdapter();
+ * $sessionPool = new CacheSessionPool($cache);
+ *
+ * $database = $spanner->connect('my-instance', 'my-database', [
+ *     'sessionPool' => $sessionPool
+ * ]);
+ * ```
+ */
+class CacheSessionPool implements SessionPoolInterface
+{
+    const CACHE_KEY_TEMPLATE = 'cache-session-pool.%s.%s';
+
+    /**
+     * @var array
+     */
+    private static $defaultConfig = [
+        'maxSessions' => PHP_INT_MAX,
+        'minSessions' => 1,
+        'shouldWaitForSession' => true,
+        'maxCyclesToWaitForSession' => 30,
+        'sleepIntervalSeconds' => .5
+    ];
+
+    /**
+     * @var CacheItemPoolInterface
+     */
+    private $cacheItemPool;
+
+    /**
+     * @var string
+     */
+    private $cacheKey;
+
+    /**
+     * @var array
+     */
+    private $config;
+
+    /**
+     * @var Database
+     */
+    private $database;
+
+    /**
+     * @param CacheItemPoolInterface $cacheItemPool A PSR-6 compatible cache
+     *        implementation used to store the session data.
+     * @param array $config [optional] {
+     *     Configuration Options.
+     *
+     *     @type int $maxSessions The maximum number of sessions to store in the
+     *           pool. **Defaults to** PHP_INT_MAX.
+     *     @type int $minSessions The minimum number of sessions to store in the
+     *           pool. **Defaults to** `1`.
+     *     @type bool $shouldWaitForSession If the pool is full, whether to block
+     *           until a new session is available. **Defaults to* `true`.
+     *     @type int $maxCyclesToWaitForSession The maximum number cycles to
+     *           wait for a session before throwing an exception. **Defaults to**
+     *           `30`. Ignored when $shouldWaitForSession is `false`.
+     *     @type float $sleepIntervalSeconds The sleep interval between cycles.
+     *           **Defaults to** `0.5`. Ignored when $shouldWaitForSession is
+     *           `false`.
+     *     @type LockInterface $lock A lock implementation capable of blocking.
+     *           **Defaults to** an flock based implementation.
+     * }
+     * @throws \InvalidArgumentException
+     */
+    public function __construct(CacheItemPoolInterface $cacheItemPool, array $config = [])
+    {
+        $this->cacheItemPool = $cacheItemPool;
+        $this->config = $config + self::$defaultConfig;
+
+        if (!isset($this->config['lock'])) {
+            $this->config['lock'] = $this->getDefaultLock();
+        }
+
+        $this->validateConfig();
+    }
+
+    /**
+     * Acquire a session from the pool.
+     *
+     * @param string $context The type of session to fetch. May be either `r`
+     *        (READ) or `rw` (READ/WRITE). **Defaults to** `r`.
+     * @return Session
+     * @throws \RuntimeException
+     */
+    public function acquire($context = SessionPoolInterface::CONTEXT_READ)
```

Comment:
It isn't being used in this cached based implementation. Other implementations could use this when it makes sense to prime transactions.